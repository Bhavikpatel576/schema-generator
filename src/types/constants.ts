export enum DataTypeName {
  CHAR = 'char',
  VARCHAR = 'varchar',
  TEXT = 'text',
  ENUM = 'enum',

  INT = 'int',
  BIGINT = 'bigint',

  DATETIME = 'datetime',
}

export enum Constraint {
  FOREIGN_KEY = 'FOREIGN_KEY',
}
