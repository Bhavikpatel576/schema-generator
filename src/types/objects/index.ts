export { DataType } from './DataType';
export { Property } from './Property';
export { Entity } from './Entity';
