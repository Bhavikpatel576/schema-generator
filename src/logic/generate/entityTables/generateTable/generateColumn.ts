import { DataTypeName, Property } from '../../../../types';

/*
  // TODO: generalize w/ adapter pattern to other languages
*/
const textTypes = [DataTypeName.CHAR, DataTypeName.VARCHAR, DataTypeName.TEXT, DataTypeName.ENUM];
export const generateColumn = ({ columnName, property }: { columnName: string, property: Property }) => {

  // define the type modifier
  let typeModifier = '';
  if (property.type.precision) typeModifier = `(${property.type.precision})`;
  if (property.type.values) typeModifier = `(${property.type.values.map(opt => "'" + opt + "'").join(',')})`; // tslint:disable-line prefer-template

  // define default over-rides
  let defaultNullRelation = (property.nullable) ? 'DEFAULT NULL' : ''; // if nullable, then the "default" default is "null"; else, nothing
  if (property.type.name === DataTypeName.TEXT) defaultNullRelation = ''; // if its varchar, then even if nullable don't show default as null; // TODO: discover why SHOW CREATE has this exception

  // define the full column definition
  const modifiers = [
    `\`${columnName}\``,

    `${property.type.name}${typeModifier}`, // empty string if precision not defined

    (textTypes.includes(property.type.name)) ? 'COLLATE utf8mb4_unicode_ci' : '', // coalate if its a string type;

    (property.nullable) ? '' : 'NOT NULL', // if not nullable, then NOT NULL

    (property.default) ? `DEFAULT ${property.default}` : defaultNullRelation, // if default is set, use it; otherwise, no default

    (property.comment) ? `COMMENT '${property.comment}'` : '', // empty string if comment not defined
  ];
  return modifiers
    .filter(modifier => !!modifier) // filter out null
    .join(' '); // merge into column
};
