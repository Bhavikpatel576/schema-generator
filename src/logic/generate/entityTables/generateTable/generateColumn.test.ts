import { DataType, DataTypeName, Property } from '../../../../types';
import { generateColumn } from './generateColumn';

describe('generateColumn', () => {
  it('should define column name accurately', () => {
    const property = new Property({
      type: new DataType({
        name: DataTypeName.INT,
      }),
    });
    const sql = generateColumn({ columnName: 'user_id', property });
    expect(sql).toContain('`user_id`');
    expect(sql).toMatchSnapshot(); // to log an example, not to actualy test logic
  });
  it('should define type name accurately', () => {
    const property = new Property({
      type: new DataType({
        name: DataTypeName.INT,
      }),
    });
    const sql = generateColumn({ columnName: 'user_id', property });
    expect(sql).toContain('int');
  });
  it('should define precision accurately, when specified', () => {
    const property = new Property({
      type: new DataType({
        name: DataTypeName.INT,
        precision: 11,
      }),
    });
    const sql = generateColumn({ columnName: 'user_id', property });
    expect(sql).toContain('int(11)');
    expect(sql).toMatchSnapshot(); // to log an example, not to actualy test logic
  });
  it('should specify NOT NULL by default', () => {
    const property = new Property({
      type: new DataType({
        name: DataTypeName.INT,
      }),
    });
    const sql = generateColumn({ columnName: 'user_id', property });
    expect(sql).toContain('NOT NULL');
    expect(sql).toMatchSnapshot(); // to log an example, not to actualy test logic
  });
  it('should be able to specify a comment', () => {
    const property = new Property({
      type: new DataType({
        name: DataTypeName.INT,
      }),
      comment: 'hey there',
    });
    const sql = generateColumn({ columnName: 'user_id', property });
    expect(sql).toContain("COMMENT 'hey there'");
    expect(sql).toMatchSnapshot(); // to log an example, not to actualy test logic
  });
  it('should be able to specify a default value', () => {
    const property = new Property({
      type: new DataType({
        name: DataTypeName.DATETIME,
        precision: 6,
      }),
      default: 'CURRENT_TIMESTAMP(6)',
    });
    const sql = generateColumn({ columnName: 'user_id', property });
    expect(sql).toContain('DEFAULT CURRENT_TIMESTAMP(6)');
    expect(sql).toMatchSnapshot(); // to log an example, not to actualy test logic
  });
  it('should be able to specify a default value while nullable', () => {
    const property = new Property({
      type: new DataType({
        name: DataTypeName.ENUM,
        values: ['option_one', 'option_two'],
      }),
      nullable: true,
      default: 'CURRENT_TIMESTAMP(6)',
    });
    const sql = generateColumn({ columnName: 'user_id', property });
    expect(sql).toContain('DEFAULT CURRENT_TIMESTAMP(6)');
    expect(sql).toMatchSnapshot(); // to log an example, not to actualy test logic
  });
  it('should be able to specify an enum column', () => {

  });
});
