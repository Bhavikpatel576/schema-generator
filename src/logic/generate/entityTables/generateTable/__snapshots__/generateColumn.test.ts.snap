// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateColumn should be able to specify a comment 1`] = `"\`user_id\` int NOT NULL COMMENT 'hey there'"`;

exports[`generateColumn should be able to specify a default value 1`] = `"\`user_id\` datetime(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6)"`;

exports[`generateColumn should be able to specify a default value while nullable 1`] = `"\`user_id\` enum('option_one','option_two') COLLATE utf8mb4_unicode_ci DEFAULT CURRENT_TIMESTAMP(6)"`;

exports[`generateColumn should define column name accurately 1`] = `"\`user_id\` int NOT NULL"`;

exports[`generateColumn should define precision accurately, when specified 1`] = `"\`user_id\` int(11) NOT NULL"`;

exports[`generateColumn should specify NOT NULL by default 1`] = `"\`user_id\` int NOT NULL"`;
